blob/ribs_store.go

```diff
@@ -0,0 +1,253 @@
+package blob
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/ipfs/boxo/chunker"
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	"github.com/lotus-web3/ribs"
+	"github.com/lotus-web3/ribs/rbdeal"
+	"github.com/multiformats/go-multihash"
+)
+
+// TODO parameterize this.
+const ribsStoreChunkSize = 4 << 20 // 4 MiB
+
+var (
+	_ Store             = (*RibsStore)(nil)
+	_ io.ReadSeekCloser = (*ribsStoredBlobReader)(nil)
+)
+
+type (
+	// RibsStore is an experimental Store implementation that uses RIBS.
+	// See: https://github.com/filcat/ribs
+	RibsStore struct {
+		ribs    ribs.RIBS
+		maxSize int
+		//index   map[uuid.UUID]*ribsStoredBlob // TODO persist this on disk
+		indexDir string
+	}
+	ribsStoredBlob struct {
+		*Descriptor
+		Chunks []cid.Cid `json:"chunks"`
+	}
+	ribsStoredBlobReader struct {
+		sess   ribs.Session
+		blob   *ribsStoredBlob
+		offset int64
+
+		currentChunkIndex       int
+		currentChunkReader      *bytes.Reader
+		currentChunkPendingSeek int64
+	}
+)
+
+// NewRibsStore instantiates a new experimental RIBS store.
+func NewRibsStore(dir string) (*RibsStore, error) {
+	rbdealDir := path.Join(path.Clean(dir), "rbdeal")
+	if err := os.Mkdir(rbdealDir, 0750); err != nil {
+		return nil, fmt.Errorf("failed to create internal directories: %w", err)
+	}
+	indexDir := path.Join(path.Clean(dir), "index")
+	if err := os.Mkdir(indexDir, 0750); err != nil {
+		return nil, fmt.Errorf("failed to create internal directories: %w", err)
+	}
+
+	// TODO Path to wallet is hardcoded in RIBS. Parameterise it and allow user to configure
+	//      See: https://github.com/FILCAT/ribs/blob/7c8766206ec1e5ec30c613cde2b3a49d0ccf25d0/rbdeal/ribs.go#L156
+
+	rbs, err := rbdeal.Open(rbdealDir)
+	if err != nil {
+		return nil, err
+	}
+	return &RibsStore{
+		ribs:    rbs,
+		maxSize: 32 << 30, // 32 GiB
+		//index:   map[uuid.UUID]*ribsStoredBlob{},
+		indexDir: indexDir,
+	}, nil
+
+}
+
+func (r *RibsStore) Start(_ context.Context) error {
```

Comment:
I believe this is not currently called in this PR? See change in #29 (to call store.Start/Shutdown from within Motion.Start/Shutdown